man/mhlogin.man
.TH MHLOGIN %manext1% 2016-10-10 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
mhlogin \- nmh login to external (OAuth) services
.SH SYNOPSIS
.HP 5
.na
.B mhlogin
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-user
.IR username ]
.RB [ \-saslmech
.IR mechanism ]
.RB [ \-authservice
.IR service ]
.RB [ \-browser
.IR command ]
.RB [ \-snoop ]
.ad
.SH DESCRIPTION
.B mhlogin
currently only supports OAuth for Gmail.  Run
.B mhlogin
.B \-user
.I username
.B \-saslmech xoauth2
.B \-authservice
.I gmail
and load the printed URL in your browser.  Login to a Gmail account, grant
authorization, and copy and paste the code into the
.B mhlogin
prompt.  Be sure to use the same account with the
.B -user
switch to
.BR send .
See the EXAMPLES section below for an example.
.PP
The
.B \-browser
switch causes
.B mhlogin
to load the URL directly into a new browser session.  The command argument
must include the browser invocation name, and can include any browser arguments
in the
.B \-browser
argument string, e.g., 'google-chrome --new-window'.
.PP
The
.B \-snoop
switch can be used to view the HTTP transaction.
.PP
All parameters configuring the service may be overridden by profile components,
and even though only Gmail is supported out of the box, the user can define
new services entirely in the profile.  Profile components are prefixed by
.IR oauth-service- ,
for example,
.I oauth-gmail-credential-file
which specifies where
.B mhlogin
should write credentials and where
.B send
should read them.
.SH "EXAMPLES"
.PP
.SS Gmail
First, login to the Gmail account that you want to use, e.g, user@example.com,
using a web browser.
Then, enter a
.B mhlogin
command such as:
.PP
.RS
.nf
mhlogin -user user@example.com -saslmech xoauth2 -authservice gmail \
-browser 'google-chrome --new-window'
.fi
.RE
.PP
The
.I username
should be a complete email address.
Respond to the prompt in the browser window, then copy and paste the
authorization code in response to the
.B mhlogin
prompt.
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName          'u
^oauth-gmail-credential-file:~^oauth-gmail
^oauth-gmail-client_id:~^nmh project client_id
^oauth-gmail-client_secret:~^nmh project client_secret
^oauth-gmail-auth_endpoint:~^https://accounts.google.com/o/oauth2/auth
^oauth-gmail-redirect_uri:~^urn:ietf:wg:oauth:2.0:oob
^oauth-gmail-token_endpoint:~^https://accounts.google.com/o/oauth2/token
^oauth-gmail-scope:~^https://mail.google.com/
.fi
.SH "SEE ALSO"
.IR send (1)
.SH DEFAULTS
.nf
.RB ` \-saslmech "' xoauth2"
.fi


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/mhlogin.c
/*
 * mhlogin.c -- login to external (OAuth) services
 *
 * This code is Copyright (c) 2014, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <errno.h>
#include <stdio.h>
#include <string.h>

#include <h/mh.h>
#include <h/utils.h>
#include <h/oauth.h>

#define MHLOGIN_SWITCHES \
    X("user username", 0, USERSW) \
    X("saslmech", 0, SASLMECHSW) \
    X("authservice", 0, AUTHSERVICESW) \
    X("browser", 0, BROWSERSW) \
    X("snoop", 0, SNOOPSW) \
    X("help", 0, HELPSW) \
    X("version", 0, VERSIONSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MHLOGIN);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MHLOGIN, switches);
#undef X

#ifdef OAUTH_SUPPORT
/* XXX copied from install-mh.c */
static char *
geta (void)
{
    static char line[BUFSIZ];

    if (fgets(line, sizeof(line), stdin) == NULL)
	done (1);
    trim_suffix_c(line, '\n');

    return line;
}

static int
do_login(const char *svc, const char *user, const char *browser, int snoop)
{
    char *fn, *code;
    mh_oauth_ctx *ctx;
    mh_oauth_cred *cred;
    FILE *cred_file;
    int failed_to_lock = 0;
    const char *url;

    if (svc == NULL) {
        adios(NULL, "missing -authservice switch");
    }

    if (user == NULL) {
        adios(NULL, "missing -user switch");
    }

    if (!mh_oauth_new(&ctx, svc)) {
        adios(NULL, mh_oauth_get_err_string(ctx));
    }

    if (snoop) {
        mh_oauth_log_to(stderr, ctx);
    }

    fn = mh_xstrdup(mh_oauth_cred_fn(svc));

    if ((url = mh_oauth_get_authorize_url(ctx)) == NULL) {
      adios(NULL, mh_oauth_get_err_string(ctx));
    }

    if (browser) {
        char *command = concat(browser, " '", url, "'", NULL);
        int status = OK;

        printf("Follow the prompts in your browser to authorize nmh"
               " to access %s.\n",
               mh_oauth_svc_display_name(ctx));
        sleep(1);

        status = system(command);
        free(command);

        if (status != OK) {
            adios ((char *) browser, "SYSTEM");
        }
    } else {
        printf("Load the following URL in your browser and authorize nmh"
               " to access %s:\n\n%s\n\n",
               mh_oauth_svc_display_name(ctx), url);
    }
    printf("Enter the authorization code: ");
    fflush(stdout);
    code = geta();

    while (!*code ||
           ((cred = mh_oauth_authorize(code, ctx)) == NULL
            && mh_oauth_get_err_code(ctx) == MH_OAUTH_BAD_GRANT)) {
      printf(!*code  ?  "Empty code; try again? "  :  "Code rejected; try again? ");
      fflush(stdout);
      code = geta();
    }
    if (cred == NULL) {
      inform("error exchanging code for OAuth2 token");
      adios(NULL, mh_oauth_get_err_string(ctx));
    }

    cred_file = lkfopendata(fn, "r+", &failed_to_lock);
    if (cred_file == NULL && errno == ENOENT) {
        cred_file = lkfopendata(fn, "w+", &failed_to_lock);
    }
    if (cred_file == NULL || failed_to_lock) {
      adios(fn, "oops");
    }
    if (!mh_oauth_cred_save(cred_file, cred, user)) {
      adios(NULL, mh_oauth_get_err_string(ctx));
    }
    if (lkfclosedata(cred_file, fn) != 0) {
      adios (fn, "oops");
    }

    free(fn);
    mh_oauth_cred_free(cred);
    mh_oauth_free(ctx);

    return 0;
}
#endif

int
main(int argc, char **argv)
{
    char *cp, **argp, **arguments;
    const char *user = NULL, *saslmech = NULL, *svc = NULL, *browser = NULL;
    int snoop = 0;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
            char help[BUFSIZ];
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW:
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW:
		adios (NULL, "-%s unknown", cp);

	    case HELPSW:
		snprintf(help, sizeof(help), "%s [switches]",
                         invo_name);
		print_help (help, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case USERSW:
		if (!(user = *argp++) || *user == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    case SASLMECHSW:
		if (!(saslmech = *argp++) || *saslmech == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    case AUTHSERVICESW:
                if (!(svc = *argp++) || *svc == '-')
                    adios (NULL, "missing argument to %s", argp[-2]);
                continue;

	    case BROWSERSW:
                if (!(browser = *argp++) || *browser == '-')
                    adios (NULL, "missing argument to %s", argp[-2]);
                continue;

            case SNOOPSW:
                snoop++;
                continue;
	    }
	}
        adios(NULL, "extraneous arguments");
    }

    if (saslmech  &&  strcasecmp(saslmech, "xoauth2")) {
        /* xoauth is assumed */
        adios(NULL, "only -saslmech xoauth2 is supported");
    }
    free(arguments);

#ifdef OAUTH_SUPPORT
    return do_login(svc, user, browser, snoop);
#else
    NMH_UNUSED(svc);
    NMH_UNUSED(browser);
    NMH_UNUSED(snoop);
    adios(NULL, "not built with OAuth support");
    return 1;
#endif
}
